substrate/client/authority-discovery/src/worker.rs

```diff
@@ -373,54 +369,101 @@ where
+				// while `is_global()` in the standard library currently isn't.
+				multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() => false,
+				multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() => false,
+				_ => true,
+			})
+		};
+
+		// These are the addresses the node is listening for incoming connections,
+		// as reported by installed protocols (tcp / websocket etc).
+		//
+		// We double check the address is global. In other words, we double check the node
+		// is not running behind a NAT.
+		// Note: we do this regardless of the `publish_non_global_ips` setting, since the
+		// node discovers many external addresses via the identify protocol.
+		let mut global_listen_addresses = self
+			.network
+			.listen_addresses()
+			.into_iter()
+			.filter_map(|address| {
+				if address_is_global(&address) {
+					Some(address_without_p2p(address, local_peer_id))
 				} else {
-					Some(address)
+					None
 				}
-			}))
-			.filter(move |address| {
+			})
+			.take(MAX_GLOBAL_LISTEN_ADDRESSES)
+			.peekable();
+
+		// Similar to listen addresses that takes into consideration `publish_non_global_ips`.
+		let mut external_addresses = self
+			.network
+			.external_addresses()
+			.into_iter()
+			.filter_map(|address| {
+				let address = address_without_p2p(address, local_peer_id);
 				if publish_non_global_ips {
-					return true
+					Some(address)
+				} else if address_is_global(&address) {
+					Some(address)
+				} else {
+					None
 				}
-
-				address.iter().all(|protocol| match protocol {
-					// The `ip_network` library is used because its `is_global()` method is stable,
-					// while `is_global()` in the standard library currently isn't.
-					multiaddr::Protocol::Ip4(ip) if !IpNetwork::from(ip).is_global() => false,
-					multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() => false,
-					_ => true,
-				})
 			})
+			.peekable();
+
+		let has_global_listen_addresses = global_listen_addresses.peek().is_some();
+		trace!(
+			target: LOG_TARGET,
+			"Node has public addresses: {}, global listen addresses: {}, external addresses: {}",
+			!self.public_addresses.is_empty(),
+			has_global_listen_addresses,
+			external_addresses.peek().is_some(),
+		);
+
+		let mut seen_addresses = HashSet::new();
+
+		let addresses = self
+			.public_addresses
+			.clone()
+			.into_iter()
+			.chain(global_listen_addresses)
+			.chain(external_addresses)
+			// Deduplicate addresses.
+			.filter(|address| seen_addresses.insert(address.clone()))
+			.take(MAX_ADDRESSES_TO_PUBLISH)
 			.collect::<Vec<_>>();
 
 		if !addresses.is_empty() {
 			debug!(
 				target: LOG_TARGET,
 				"Publishing authority DHT record peer_id='{local_peer_id}' with addresses='{addresses:?}'",
 			);
+
+			if !self.warn_public_addresses &&
+				self.public_addresses.is_empty() &&
+				!has_global_listen_addresses
+			{
+				self.warn_public_addresses = true;
+
+				warn!(
```

Comment:
```suggestion
				error!(
```

Reply:
Makes sense, that would be a bit more descriptive, should help operators in noticing this and take action